File: projects/06-object-store/src/erasure/lrc.rs
//! Local Reconstruction Codes — cheap repair for the common case (one lost shard).
//!
//! Plain RS repairs one failure by reading **all `k`** other shards. LRC splits
//! data into `l` local groups, adds one local parity per group, plus `r` global
//! (RS-style) parities. A single lost shard repairs from **≈ `k/l`** reads.
//!
//! Lab target from docs/12 §6.1: **`(k=4, l=2, r=2)`** — 4 data + 2 local + 2
//! global = 8 shards. Repairing one data shard reads 2 (local group), not 4.
//!
//! ## Observable (SPEC)
//!
//! Measure repair-read fan-in both ways: LRC single-shard repair ≈ `k/l`, plain
//! RS = `k`. Bodies are `todo!()`; reuse [`super::reed_solomon`] / [`super::gf256`]
//! for the global parities.

use super::gf256::Gf256;
use super::reed_solomon::Shard;
use super::ErasureError;

/// Lab LRC shape constants (docs/12 §6.1): `(k=4, l=2, r=2)`.
const LAB_K: usize = 4;
const LAB_L: usize = 2;
const LAB_R: usize = 2;
/// Data shards per local group when the `k` data shards split evenly across `l`.
const GROUP_SIZE: usize = LAB_K / LAB_L; // 2

/// Global (RS-style) parity rows over the `k` data shards, docs/12 §6.1.
/// Same GF(2⁸) tables and Vandermonde flavour as [`super::reed_solomon`]'s Q row —
/// LRC reorganizes which shards a parity covers, it does not invent new math.
const GLOBAL_COEFFS: [[u8; LAB_K]; LAB_R] = [
    [1, 2, 4, 8],  // G0 = 1·d0 ⊕ 2·d1 ⊕ 4·d2 ⊕ 8·d3
    [1, 3, 9, 27], // G1 = 1·d0 ⊕ 3·d1 ⊕ 9·d2 ⊕ 27·d3
];

/// LRC parameters `(k, l, r)`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LrcParams {
    /// Data shards.
    pub k: usize,
    /// Number of local groups (one local parity each).
    pub l: usize,
    /// Global (RS-style) parity shards.
    pub r: usize,
}

impl LrcParams {
    /// Docs §6.1 lab shape: 4 data, 2 local groups, 2 global parities.
    pub const LAB_4_2_2: Self = Self { k: 4, l: 2, r: 2 };

    /// Total shards = data + local parities + global parities.
    pub fn n(self) -> usize {
        self.k + self.l + self.r
    }

    /// Ideal single-shard repair-read fan-in ≈ `k / l` (when groups are equal).
    pub fn local_repair_fan_in(self) -> usize {
        self.k / self.l
    }
}

/// How many shards a repair actually touched — the metric the SPEC wants.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct RepairStats {
    /// Shards read to rebuild the missing one.
    pub shards_read: usize,
    /// Whether repair used a local group (`true`) or fell back to globals.
    pub used_local: bool,
}

/// Local Reconstruction Code coder.
#[derive(Debug, Clone)]
pub struct Lrc {
    pub params: LrcParams,
    pub gf: Gf256,
}

impl Lrc {
    /// Lab LRC `(4, 2, 2)`.
    pub fn lab() -> Self {
        Self {
            params: LrcParams::LAB_4_2_2,
            gf: Gf256::new(),
        }
    }

    /// Encode plaintext → `n = k + l + r` shards (docs/12 §6.1).
    ///
    /// Systematic layout: ids `0..k` data, `k..k+l` local XOR parities (one per
    /// group), `k+l..k+l+r` global RS-style parities across all `k` data shards.
    ///
    /// # Errors
    ///
    /// [`ErasureError::InvalidLayout`] on empty input; [`ErasureError::Unsupported`]
    /// if this is not the lab `(4, 2, 2)` shape (the coefficient tables are fixed).
    pub fn encode(&self, plaintext: &[u8]) -> Result<Vec<Shard>, ErasureError> {
        self.assert_lab_shape()?;
        if plaintext.is_empty() {
            return Err(ErasureError::InvalidLayout(
                "plaintext must be non-empty".into(),
            ));
        }

        // Zero-pad to a multiple of k, then split into k contiguous data shards.
        let k = self.params.k;
        let mut padded = plaintext.to_vec();
        let rem = padded.len() % k;
        if rem != 0 {
            padded.resize(padded.len() + (k - rem), 0);
        }
        let shard_len = padded.len() / k;
        let data: Vec<Vec<u8>> = padded.chunks(shard_len).map(<[u8]>::to_vec).collect();

        let mut shards: Vec<Shard> = data
            .iter()
            .enumerate()
            .map(|(id, bytes)| Shard::new(id, bytes.clone()))
            .collect();

        // Local parities: pure XOR over each group's data shards (docs/12 §6.1).
        for g in 0..self.params.l {
            let members = self.group_data_ids(g);
            let mut parity = vec![0u8; shard_len];
            for &d in &members {
                for (col, byte) in parity.iter_mut().enumerate() {
                    *byte = Gf256::add(*byte, data[d][col]);
                }
            }
            shards.push(Shard::new(k + g, parity));
        }

        // Global parities: RS-style Vandermonde rows over ALL k data shards.
        for (t, coeffs) in GLOBAL_COEFFS.iter().enumerate() {
            let mut parity = vec![0u8; shard_len];
            for (byte_idx, byte) in parity.iter_mut().enumerate() {
                let mut acc = 0u8;
                for (c, chunk) in data.iter().enumerate() {
                    acc = Gf256::add(acc, self.gf.mul(coeffs[c], chunk[byte_idx]));
                }
                *byte = acc;
            }
            shards.push(Shard::new(k + self.params.l + t, parity));
        }

        Ok(shards)
    }

    /// Repair a single missing shard; return the rebuilt shard + read fan-in.
    ///
    /// Prefers the local group (`used_local = true`, `shards_read ≈ k/l`): a lost
    /// data or local-parity shard is the XOR of the rest of its group. Falls back
    /// to the global parities (`used_local = false`, `shards_read = k`) when the
    /// local group is itself incomplete, or when the missing shard *is* a global.
    ///
    /// # Errors
    ///
    /// [`ErasureError::InvalidLayout`] for a wrong slot count or an out-of-range
    /// `missing`; [`ErasureError::TooManyErasures`] if neither the local group nor
    /// the globals leave enough survivors to rebuild it.
    pub fn repair_one(
        &self,
        shards: &[Option<Shard>],
        missing: usize,
    ) -> Result<(Shard, RepairStats), ErasureError> {
        self.assert_lab_shape()?;
        let n = self.params.n();
        if shards.len() != n {
            return Err(ErasureError::InvalidLayout(format!(
                "expected {n} shard slots, got {}",
                shards.len()
            )));
        }
        if missing >= n {
            return Err(ErasureError::InvalidLayout(format!(
                "missing id {missing} out of range 0..{n}"
            )));
        }

        // Fast path: rebuild from the missing shard's local group with one XOR pass.
        if let Some(group) = self.local_group_of(missing) {
            let peers: Vec<usize> = group.into_iter().filter(|&id| id != missing).collect();
            if peers.iter().all(|&id| shards[id].is_some()) {
                let shard_len = shards[peers[0]].as_ref().unwrap().data.len();
                let mut data = vec![0u8; shard_len];
                for &id in &peers {
                    let s = shards[id].as_ref().unwrap();
                    for (col, byte) in data.iter_mut().enumerate() {
                        *byte = Gf256::add(*byte, s.data[col]);
                    }
                }
                return Ok((
                    Shard::new(missing, data),
                    RepairStats {
                        shards_read: peers.len(),
                        used_local: true,
                    },
                ));
            }
        }

        // Fallback: use a global parity row. Needs every OTHER data shard present.
        self.repair_via_global(shards, missing)
    }

    /// Global-parity fallback for a single missing data or local-parity shard.
    ///
    /// Recomputes the shard from all `k` data shards (recovering the one missing
    /// data shard first via a global row whose coefficient on it is nonzero).
    /// Reads `k` shards — the same fan-in plain RS pays, which is exactly the cost
    /// LRC's local groups avoid in the common case.
    fn repair_via_global(
        &self,
        shards: &[Option<Shard>],
        missing: usize,
    ) -> Result<(Shard, RepairStats), ErasureError> {
        let k = self.params.k;

        // Collect the k data shards, tolerating at most the one we're rebuilding.
        let mut data: Vec<Option<&Vec<u8>>> = (0..k)
            .map(|id| shards[id].as_ref().map(|s| &s.data))
            .collect();
        let missing_data: Vec<usize> = (0..k).filter(|&id| data[id].is_none()).collect();

        // More than one hole among the data shards is beyond a single-shard repair.
        let recovered_missing_data: Vec<u8>;
        if missing_data.len() > 1 || (missing_data.len() == 1 && missing_data[0] != missing) {
            return Err(ErasureError::TooManyErasures {
                need: k,
                have: k - missing_data.len(),
            });
        }
        if let Some(&hole) = missing_data.first() {
            // Recover the missing data shard d[hole] from a present global parity:
            //   G = Σ coeff[c]·d[c]  ⇒  d[hole] = (G ⊕ Σ_{c≠hole} coeff[c]·d[c]) / coeff[hole]
            let (coeffs, parity) =
                self.available_global(shards)
                    .ok_or(ErasureError::TooManyErasures {
                        need: k,
                        have: k - 1,
                    })?;
            let shard_len = parity.len();
            let coeff_hole = coeffs[hole];
            let inv = self.gf.inv(coeff_hole)?;
            let mut rebuilt = vec![0u8; shard_len];
            for (col, byte) in rebuilt.iter_mut().enumerate() {
                let mut acc = parity[col];
                for c in 0..k {
                    if c == hole {
                        continue;
                    }
                    acc = Gf256::add(acc, self.gf.mul(coeffs[c], data[c].unwrap()[col]));
                }
                *byte = self.gf.mul(acc, inv);
            }
            recovered_missing_data = rebuilt;
            data[hole] = Some(&recovered_missing_data);
        }

        // Every data shard is now known; recompute whichever shard was asked for.
        let shard_bytes = self.recompute_shard(missing, &data);
        Ok((
            Shard::new(missing, shard_bytes),
            RepairStats {
                shards_read: k,
                used_local: false,
            },
        ))
    }

    /// The first present global parity as `(coefficients, bytes)`, if any.
    fn available_global<'a>(
        &self,
        shards: &'a [Option<Shard>],
    ) -> Option<(&'static [u8; LAB_K], &'a Vec<u8>)> {
        let base = self.params.k + self.params.l;
        (0..self.params.r).find_map(|t| {
            shards[base + t]
                .as_ref()
                .map(|s| (&GLOBAL_COEFFS[t], &s.data))
        })
    }

    /// Recompute shard `id` from the full set of `k` data shards.
    fn recompute_shard(&self, id: usize, data: &[Option<&Vec<u8>>]) -> Vec<u8> {
        let k = self.params.k;
        let shard_len = data.iter().flatten().next().map_or(0, |d| d.len());
        if id < k {
            return data[id].unwrap().clone();
        }
        if id < k + self.params.l {
            let g = id - k;
            let mut parity = vec![0u8; shard_len];
            for &d in &self.group_data_ids(g) {
                for (col, byte) in parity.iter_mut().enumerate() {
                    *byte = Gf256::add(*byte, data[d].unwrap()[col]);
                }
            }
            return parity;
        }
        let coeffs = &GLOBAL_COEFFS[id - k - self.params.l];
        let mut parity = vec![0u8; shard_len];
        for (col, byte) in parity.iter_mut().enumerate() {
            let mut acc = 0u8;
            for (c, d) in data.iter().enumerate() {
                acc = Gf256::add(acc, self.gf.mul(coeffs[c], d.unwrap()[col]));
            }
            *byte = acc;
        }
        parity
    }

    /// Data shard ids belonging to local group `g` (equal-sized groups).
    fn group_data_ids(&self, g: usize) -> Vec<usize> {
        let start = g * GROUP_SIZE;
        (start..start + GROUP_SIZE).collect()
    }

    /// Ids in the local group that covers shard `id`: its data shards + local
    /// parity. `None` for a global parity (globals have no local group).
    fn local_group_of(&self, id: usize) -> Option<Vec<usize>> {
        let k = self.params.k;
        let g = if id < k {
            id / GROUP_SIZE
        } else if id < k + self.params.l {
            id - k
        } else {
            return None;
        };
        let mut members = self.group_data_ids(g);
        members.push(k + g); // the group's local parity
        Some(members)
    }

    /// The coefficient tables are hard-wired to `(4, 2, 2)`, mirroring how
    /// [`super::reed_solomon`] asserts its RS(4,2) shape.
    fn assert_lab_shape(&self) -> Result<(), ErasureError> {
        let p = self.params;
        if (p.k, p.l, p.r) != (LAB_K, LAB_L, LAB_R) {
            return Err(ErasureError::Unsupported(format!(
                "only lab LRC(4,2,2) is implemented, got ({}, {}, {})",
                p.k, p.l, p.r
            )));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lab_params_fan_in() {
        let p = LrcParams::LAB_4_2_2;
        assert_eq!(p.n(), 8);
        assert_eq!(p.local_repair_fan_in(), 2); // k/l = 4/2
    }

    #[test]
    fn single_data_shard_repairs_with_local_fan_in() {
        let lrc = Lrc::lab();
        let plaintext: Vec<u8> = (0u8..32).collect();
        let shards = lrc.encode(&plaintext).expect("encode");
        let missing = 0usize; // d0 in group A
        let mut present: Vec<Option<Shard>> = shards.into_iter().map(Some).collect();
        present[missing] = None;
        let (rebuilt, stats) = lrc.repair_one(&present, missing).expect("repair");
        assert_eq!(rebuilt.id, missing);
        assert!(stats.used_local);
        assert_eq!(
            stats.shards_read,
            lrc.params.local_repair_fan_in(),
            "SPEC: single-shard repair reads ≈ k/l, not k"
        );
    }
}

File: projects/06-object-store/src/object_store/erasure/lrc.py
"""Local Reconstruction Codes — cheap repair for the common case, one lost shard.

Plain RS repairs a single failure by reading **all `k`** other shards. That is
the dominant cost at scale, because single-disk failure is not the rare case —
it is essentially the *only* case, happening constantly across a large fleet
while double failures are exceptional. Rebuilding one 4 TB disk by reading 17
others saturates the network for hours, and during that window the array is
degraded.

LRC splits the data into `l` **local groups**, gives each a local XOR parity, and
adds `r` global parities on top. A single lost shard then repairs from `≈ k/l`
reads instead of `k`. Azure's published numbers are the canonical example: LRC
cut repair traffic roughly in half at a small storage-overhead increase, which
is a better trade than it sounds because repair bandwidth is the scarce resource.

Lab shape: **(k=4, l=2, r=2)** — 4 data + 2 local + 2 global = 8 shards.
Repairing one data shard reads 2, not 4.

## The catch worth understanding

LRC is not simply "better RS". Adding local parities costs storage, and the code
is no longer maximum-distance-separable: with `k + l + r` shards it does *not*
tolerate every combination of `l + r` losses the way an MDS code of the same
width would. What it buys is that the failure you actually see, constantly, gets
much cheaper. That is a deliberate trade of worst-case tolerance for common-case
cost.
"""

from __future__ import annotations

from dataclasses import dataclass

from . import InvalidLayout, TooManyErasures, Unsupported
from .gf256 import Gf256
from .reed_solomon import Shard

__all__ = ["Lrc", "LrcParams", "RepairStats"]

LAB_K = 4
LAB_L = 2
LAB_R = 2
GROUP_SIZE = LAB_K // LAB_L

GLOBAL_COEFFICIENTS = (
    (1, 2, 4, 8),
    (1, 3, 9, 27),
)
"""Two Vandermonde rows over all `k` data shards. Same GF(2⁸) flavour as RS's Q
row — LRC reorganises *which shards a parity covers*, it does not invent new
arithmetic. Different bases (2 and 3) keep the two rows independent."""


@dataclass(frozen=True, slots=True)
class LrcParams:
    """The shape `(k, l, r)`."""

    k: int
    l: int  # noqa: E741 - the literature's name for the local-group count
    r: int

    @property
    def n(self) -> int:
        return self.k + self.l + self.r

    @property
    def local_repair_fan_in(self) -> int:
        """Ideal single-shard repair reads, `k / l`, for equal groups."""
        return self.k // self.l


LAB_PARAMS = LrcParams(k=LAB_K, l=LAB_L, r=LAB_R)


@dataclass(frozen=True, slots=True)
class RepairStats:
    """How many shards a repair actually touched — the number the SPEC wants.

    Reported rather than assumed, because the whole claim of LRC is a *measured*
    reduction in fan-in and "it should read about two" is not evidence.
    """

    shards_read: int
    used_local: bool


class Lrc:
    """A Local Reconstruction Code coder."""

    __slots__ = ("gf", "params")

    def __init__(self, params: LrcParams = LAB_PARAMS) -> None:
        if (params.k, params.l, params.r) != (LAB_K, LAB_L, LAB_R):
            raise Unsupported(
                f"only lab LRC({LAB_K},{LAB_L},{LAB_R}) is implemented, "
                f"got ({params.k}, {params.l}, {params.r})"
            )
        self.params = params
        self.gf = Gf256()

    @classmethod
    def lab(cls) -> Lrc:
        return cls(LAB_PARAMS)

    def encode(self, plaintext: bytes) -> list[Shard]:
        """Encode into `n = k + l + r` shards.

        Systematic layout: ids `0..k` are data, `k..k+l` are the local XOR
        parities (one per group), and `k+l..n` are the global parities across
        all `k` data shards.
        """
        if not plaintext:
            raise InvalidLayout("plaintext must be non-empty")

        k = self.params.k
        padded = bytearray(plaintext)
        remainder = len(padded) % k
        if remainder:
            padded.extend(b"\x00" * (k - remainder))

        shard_len = len(padded) // k
        data = [bytes(padded[i * shard_len : (i + 1) * shard_len]) for i in range(k)]
        shards = [Shard(index, chunk) for index, chunk in enumerate(data)]

        # Local parities: plain XOR over each group. No field multiply needed —
        # that is exactly why a local repair is cheap in CPU as well as I/O.
        for group in range(self.params.l):
            parity = bytearray(shard_len)
            for member in self.group_data_ids(group):
                for column in range(shard_len):
                    parity[column] ^= data[member][column]
            shards.append(Shard(k + group, bytes(parity)))

        # Global parities: Vandermonde rows over every data shard.
        for offset, coefficients in enumerate(GLOBAL_COEFFICIENTS):
            parity = bytearray(shard_len)
            for column in range(shard_len):
                accumulator = 0
                for index, chunk in enumerate(data):
                    accumulator ^= self.gf.mul(coefficients[index], chunk[column])
                parity[column] = accumulator
            shards.append(Shard(k + self.params.l + offset, bytes(parity)))

        return shards

    def repair_one(self, shards: list[Shard | None], missing: int) -> tuple[Shard, RepairStats]:
        """Repair one missing shard, reporting the read fan-in it cost.

        Tries the local group first: a lost data or local-parity shard is simply
        the XOR of the rest of its group, so the repair is `k/l` reads and no
        multiplication. Falls back to the global parities — `k` reads, the same
        cost plain RS always pays — when the local group is itself incomplete,
        or when the missing shard *is* a global and has no local group.
        """
        n = self.params.n
        if len(shards) != n:
            raise InvalidLayout(f"expected {n} shard slots, got {len(shards)}")
        if not 0 <= missing < n:
            raise InvalidLayout(f"missing id {missing} out of range 0..{n}")

        group = self.local_group_of(missing)
        if group is not None:
            peers = [member for member in group if member != missing]
            present = [shards[member] for member in peers]
            if not any(peer is None for peer in present):
                intact = [_require(peer) for peer in present]
                shard_len = len(intact[0].data)
                data = bytearray(shard_len)
                for peer in intact:
                    for column in range(shard_len):
                        data[column] ^= peer.data[column]
                return (
                    Shard(missing, bytes(data)),
                    RepairStats(shards_read=len(intact), used_local=True),
                )

        return self._repair_via_global(shards, missing)

    def _repair_via_global(
        self, shards: list[Shard | None], missing: int
    ) -> tuple[Shard, RepairStats]:
        """Rebuild through a global parity row — the expensive path."""
        k = self.params.k
        data: list[bytes | None] = [
            shard.data if (shard := shards[index]) is not None else None for index in range(k)
        ]
        holes = [index for index in range(k) if data[index] is None]

        if len(holes) > 1 or (holes and holes[0] != missing):
            raise TooManyErasures(need=k, have=k - len(holes))

        if holes:
            hole = holes[0]
            # From `G = Σ coeff[c]·d[c]`, solve for the one unknown:
            #   d[hole] = (G ⊕ Σ_{c≠hole} coeff[c]·d[c]) / coeff[hole]
            # This is exactly why the coefficients must all be nonzero — a zero
            # coefficient on the missing column makes it unsolvable from that
            # row.
            resolved = self._available_global(shards)
            if resolved is None:
                raise TooManyErasures(need=k, have=k - 1)
            coefficients, parity = resolved

            inverse = self.gf.inv(coefficients[hole])
            rebuilt = bytearray(len(parity))
            for column in range(len(parity)):
                accumulator = parity[column]
                for index in range(k):
                    if index == hole:
                        continue
                    known = data[index]
                    assert known is not None
                    accumulator ^= self.gf.mul(coefficients[index], known[column])
                rebuilt[column] = self.gf.mul(accumulator, inverse)
            data[hole] = bytes(rebuilt)

        return (
            Shard(missing, self._recompute(missing, data)),
            RepairStats(shards_read=k, used_local=False),
        )

    def _available_global(self, shards: list[Shard | None]) -> tuple[tuple[int, ...], bytes] | None:
        """The first surviving global parity as `(coefficients, bytes)`."""
        base = self.params.k + self.params.l
        for offset in range(self.params.r):
            shard = shards[base + offset]
            if shard is not None:
                return GLOBAL_COEFFICIENTS[offset], shard.data
        return None

    def _recompute(self, shard_id: int, data: list[bytes | None]) -> bytes:
        """Recompute any shard from a complete set of data shards."""
        k = self.params.k
        if shard_id < k:
            known = data[shard_id]
            assert known is not None
            return known

        shard_len = next(len(chunk) for chunk in data if chunk is not None)

        if shard_id < k + self.params.l:
            parity = bytearray(shard_len)
            for member in self.group_data_ids(shard_id - k):
                known = data[member]
                assert known is not None
                for column in range(shard_len):
                    parity[column] ^= known[column]
            return bytes(parity)

        coefficients = GLOBAL_COEFFICIENTS[shard_id - k - self.params.l]
        parity = bytearray(shard_len)
        for column in range(shard_len):
            accumulator = 0
            for index, chunk in enumerate(data):
                assert chunk is not None
                accumulator ^= self.gf.mul(coefficients[index], chunk[column])
            parity[column] = accumulator
        return bytes(parity)

    def group_data_ids(self, group: int) -> list[int]:
        """Data shard ids in local group `group`, for equal-sized groups."""
        start = group * GROUP_SIZE
        return list(range(start, start + GROUP_SIZE))

    def local_group_of(self, shard_id: int) -> list[int] | None:
        """The local group covering `shard_id` — its data shards plus its parity.

        `None` for a global parity: globals span every group and therefore
        belong to none, which is why losing one always costs the full `k` reads.
        """
        k = self.params.k
        if shard_id < k:
            group = shard_id // GROUP_SIZE
        elif shard_id < k + self.params.l:
            group = shard_id - k
        else:
            return None
        return [*self.group_data_ids(group), k + group]


def _require(shard: Shard | None) -> Shard:
    if shard is None:  # pragma: no cover - guarded by the caller
        raise InvalidLayout("expected a present shard")
    return shard
